src/core.jl

```diff
@@ -327,178 +324,125 @@ isarray(juliaclass::String) = endswith(juliaclass, "[]")
-    result = callmethod(Ptr(obj), jfieldID)
-    result==C_NULL && geterror()
-    return convert_result(fieldType, result)
-end
-
-#Generate these methods to satisfy ccall's compile time constant requirement
-#_jcall for primitive and Nothing return types
-for (x, y, z) in [(:jboolean, :(JNI.CallBooleanMethodA), :(JNI.CallStaticBooleanMethodA)),
-                  (:jchar,    :(JNI.CallCharMethodA),    :(JNI.CallStaticCharMethodA))   ,
-                  (:jbyte,    :(JNI.CallByteMethodA),    :(JNI.CallStaticByteMethodA))   ,
-                  (:jshort,   :(JNI.CallShortMethodA),   :(JNI.CallStaticShortMethodA))  ,
-                  (:jint,     :(JNI.CallIntMethodA),     :(JNI.CallStaticIntMethodA))    ,
-                  (:jlong,    :(JNI.CallLongMethodA),    :(JNI.CallStaticLongMethodA))   ,
-                  (:jfloat,   :(JNI.CallFloatMethodA),   :(JNI.CallStaticFloatMethodA))  ,
-                  (:jdouble,  :(JNI.CallDoubleMethodA),  :(JNI.CallStaticDoubleMethodA)) ,
-                  (:jvoid,    :(JNI.CallVoidMethodA),    :(JNI.CallStaticVoidMethodA))   ]
-    m = quote
-        function _jcall(obj, jmethodId::Ptr{Nothing}, callmethod::Ptr{Nothing}, rettype::Type{$(x)},
-                        argtypes::Tuple, args...)
-            if callmethod == C_NULL #!
-                callmethod = ifelse( typeof(obj)<:JavaObject, $y , $z )
+for (x, name) in [(:Type,             "Object"),
+                  (:(Type{jboolean}), "Boolean"),
+                  (:(Type{jchar}),    "Char"   ),
+                  (:(Type{jbyte}),    "Byte"   ),
+                  (:(Type{jshort}),   "Short"  ),
+                  (:(Type{jint}),     "Int"    ),
+                  (:(Type{jlong}),    "Long"   ),
+                  (:(Type{jfloat}),   "Float"  ),
+                  (:(Type{jdouble}),  "Double" ),
+                  (:(Type{jvoid}),    "Void"   )]
+    for (t, cstr, fstr) in [(:JavaObject,    "Call$(name)MethodA",      "Get$(name)Field"),
+                            (:JavaMetaClass, "CallStatic$(name)MethodA", "GetStatic$(name)Field")]
+        callmethod = :(JNI.$(Symbol(cstr)))
+        fieldmethod = :(JNI.$(Symbol(fstr)))
+        m = quote
+            function _jfield(obj::T, jfieldID::Ptr{Nothing}, fieldType::$x) where T <: $t
+                result = $fieldmethod(Ptr(obj), jfieldID)
+                geterror()
+                return convert_result(fieldType, result)
             end
-            @assert callmethod != C_NULL
-            @assert jmethodId != C_NULL
-            isnull(obj) && throw(JavaCallError("Attempt to call method on Java NULL"))
-            savedArgs, convertedArgs = convert_args(argtypes, args...)
-            GC.@preserve savedArgs begin
-                result = callmethod(Ptr(obj), jmethodId, Array{JNI.jvalue}(jvalue.(convertedArgs)))
+            function _jcall(obj::T, jmethodId::Ptr{Nothing}, rettype::$x,
+                            argtypes::Tuple, args...; callmethod=$callmethod) where T <: $t
```

Comment:
We don't need to check for `C_NULL` as was done previously, `C_NULL` was used as a special value to be able to set a default, much like a keyword argument or default argument.

The `callmethod` kwarg is only used by `jnew`. I think this is ok, alternatively, we could break it up into two functions (manually generate functions with defaults like what julia generates for us now with kwargs). This could be more performant. There is if I recall correctly a slight performance penalty to functions with keyword arguments.